nixos/modules/services/web-apps/photoprism.nix

```diff
@@ -0,0 +1,291 @@
+      type = types.path;
+      example = "/mnt/nas/photos";
+      description = "Path to your photos.";
+    };
+
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/photoprism";
+      description = "Data directory for photoprism.";
+    };
+
+    siteTitle = mkOption {
+      type = types.str;
+      default = "PhotoPrism";
+      description = "Web UI title.";
+    };
+
+    siteCaption = mkOption {
+      type = types.str;
+      default = "Browse Your Life";
+      description = "Web UI caption.";
+    };
+
+    extraEnv = mkOption {
+      type = types.attrsOf types.str;
+      default = { };
+      example = { PHOTOPRISM_DEBUG = "true"; };
+      description = ''
+        Additional environment variables for the photoprism service.
+
+        See <link xlink:href="https://dl.photoprism.app/docker/docker-compose.yml"/>
+        and <link xlink:href="https://github.com/photoprism/photoprism/blob/develop/internal/config/flags.go"/>
+        for the supported environment variables.
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.photoprism;
+      defaultText = lib.literalExample "pkgs.photoprism";
+      description = "Photoprism package to use.";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.databaseDriver == "sqlite" -> (cfg.databaseDsn == null);
+        message = "SQLite is an internal database, databaseDsn must be null";
+      }
+      {
+        assertion = cfg.databaseDriver == "mysql" -> (cfg.databaseDsn != null);
+        message = "databaseDsn must be provided when using mysql databases";
+      }
+    ];
+
+    networking.firewall = lib.mkIf cfg.openFirewall {
+      allowedTCPPorts = [ cfg.port ];
+    };
+
+    users = {
+      users.${cfg.user} = {
+        inherit (cfg) group;
+        home = cfg.dataDir;
+        createHome = true;
+        isSystemUser = true;
+      };
+      groups.${cfg.group} = { };
+    };
+
+    systemd.services.photoprism = {
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/photoprism --assets-path ${cfg.package}/assets start
+        '';
+        User = cfg.user;
+        Group = cfg.group;
+
+        # hardening
+        DevicePolicy = "closed";
+        CapabilityBoundingSet = "";
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_UNIX" "AF_NETLINK" ];
+        DeviceAllow = "";
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
+        PrivateTmp = true;
+        PrivateUsers = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectSystem = "strict";
+        BindPaths = [
+          cfg.dataDir
+          cfg.originalsDir
```

Comment:
Does PhotoPrism need write access to that directory?

Reply:
Nope. Is there anyway we can make the access read-only?

Reply:
`BindReadOnlyPaths`